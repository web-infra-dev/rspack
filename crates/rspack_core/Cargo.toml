[package]
edition    = "2021"
license    = "MIT"
name       = "rspack_core"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

[dependencies]
anymap = "1.0.0-beta.2"
async-recursion = { workspace = true }
async-trait = { workspace = true }
bitflags = { workspace = true }
dashmap = { workspace = true }
derivative = { workspace = true }
dyn-clone = "1.0.16"
either = "1"
futures = { workspace = true }
glob-match = "0.2.1"
hashlink = { workspace = true }
hex = "0.4.3"
indexmap = { workspace = true }
itertools = { workspace = true }
json = { workspace = true }
mime_guess = { workspace = true }
once_cell = { workspace = true }
oxc_resolver = { version = "1.5.3" }
paste = { workspace = true }
petgraph = { version = "0.6.4", features = ["serde-1"] }
rayon = { workspace = true }
regex = { workspace = true }
rkyv = { workspace = true, features = ["indexmap", "validation"] }
rspack_ast = { path = "../rspack_ast" }
rspack_core_macros = { path = "../rspack_core_macros" }
rspack_database = { path = "../rspack_database", features = ["rayon"] }
rspack_error = { path = "../rspack_error" }
rspack_fs = { path = "../rspack_fs", features = ["async", "rspack-error"] }
rspack_futures = { path = "../rspack_futures" }
rspack_hash = { path = "../rspack_hash" }
rspack_hook = { path = "../rspack_hook" }
rspack_identifier = { path = "../rspack_identifier" }
rspack_loader_runner = { path = "../rspack_loader_runner" }
rspack_regex = { path = "../rspack_regex" }
rspack_sources = { workspace = true }
rspack_swc_visitors = { path = "../rspack_swc_visitors" }
rspack_util = { path = "../rspack_util" }
rustc-hash = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
sugar_path = { workspace = true }
swc_core = { workspace = true, features = [
  "__ecma",
  "__visit",
  "__utils",
  "__ecma_transforms",
  "ecma_ast",
  "ecma_codegen",
  "ecma_quote",
  "common_concurrent",
  "common_tty",
  "swc_ecma_ast",
  "ecma_transforms_react",
  "ecma_transforms_module",
  "swc_ecma_codegen",
  "swc_ecma_visit",
  "ecma_visit_path",
  "css_ast",
  "css_codegen",
  "css_parser",
  "css_utils",
  "css_visit",
  "css_visit_path",
  "css_compat",
  "css_modules",
] }
swc_node_comments = { workspace = true }
tokio = { workspace = true, features = ["rt", "rt-multi-thread", "macros", "test-util", "parking_lot"] }
tracing = { workspace = true }
url = { workspace = true }
ustr = { workspace = true }

[dev-dependencies]
pretty_assertions = { version = "1.4.0" }
