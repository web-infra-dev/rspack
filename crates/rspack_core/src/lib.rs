#![feature(let_chains)]
#![feature(if_let_guard)]
#![feature(iter_intersperse)]
#![feature(box_patterns)]
#![feature(anonymous_lifetime_in_impl_trait)]
#![feature(hash_raw_entry)]
#![feature(option_get_or_insert_default)]
#![feature(slice_group_by)]
use std::{fmt, sync::Arc};
mod dependencies_block;
pub mod diagnostics;
mod update_hash;
pub use dependencies_block::{
  AsyncDependenciesBlock, AsyncDependenciesBlockIdentifier, DependenciesBlock, DependencyLocation,
};
mod fake_namespace_object;
mod template;
pub use fake_namespace_object::*;
pub use template::Template;
mod module_profile;
pub use module_profile::*;
use rspack_database::Database;
pub mod external_module;
pub use external_module::*;
mod logger;
pub use logger::*;
pub mod cache;
mod normal_module;
mod raw_module;
pub use raw_module::*;
mod exports_info;
pub use exports_info::*;
pub mod module;
pub mod parser_and_generator;
pub use module::*;
pub use parser_and_generator::*;
mod runtime_globals;
pub use normal_module::*;
pub use runtime_globals::RuntimeGlobals;
mod plugin;
pub use plugin::*;
mod context_module;
pub use context_module::*;
mod context_module_factory;
pub use context_module_factory::*;
mod init_fragment;
pub use init_fragment::*;
mod module_factory;
pub use module_factory::*;
mod normal_module_factory;
pub use normal_module_factory::*;
mod ignore_error_module_factory;
pub use ignore_error_module_factory::*;
mod self_module_factory;
pub use self_module_factory::*;
mod self_module;
pub use self_module::*;
mod compiler;
pub use compiler::*;
mod options;
pub use options::*;
mod module_graph;
pub use module_graph::*;
mod chunk;
pub use chunk::*;
mod dependency;
pub use dependency::*;
mod utils;
use ustr::Ustr;
pub use utils::*;
mod chunk_graph;
pub use chunk_graph::*;
mod build_chunk_graph;
mod stats;
pub use stats::*;
mod runtime;
mod runtime_module;
pub use runtime::*;
pub use runtime_module::*;
mod code_generation_results;
pub use code_generation_results::*;
mod entrypoint;
pub use entrypoint::*;
mod loader;
pub use loader::*;
// mod external;
// pub use external::*;
mod chunk_group;
pub use chunk_group::*;
mod ukey;
pub use ukey::*;
pub mod resolver;
pub use resolver::*;
pub mod concatenated_module;
pub mod reserved_names;
pub mod tree_shaking;

pub use rspack_core_macros::{impl_runtime_module, impl_source_map_config};
pub use rspack_loader_runner::{get_scheme, ResourceData, Scheme, BUILTIN_LOADER_PREFIX};
pub use rspack_sources;

#[cfg(debug_assertions)]
pub mod debug_info;

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SourceType {
  JavaScript,
  Css,
  Wasm,
  Asset,
  Expose,
  Remote,
  ShareInit,
  ConsumeShared,
  #[default]
  Unknown,
}

impl std::fmt::Display for SourceType {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      SourceType::JavaScript => write!(f, "javascript"),
      SourceType::Css => write!(f, "css"),
      SourceType::Wasm => write!(f, "wasm"),
      SourceType::Asset => write!(f, "asset"),
      SourceType::Expose => write!(f, "expose"),
      SourceType::Remote => write!(f, "remote"),
      SourceType::ShareInit => write!(f, "share-init"),
      SourceType::ConsumeShared => write!(f, "consume-shared"),
      SourceType::Unknown => write!(f, "unknown"),
    }
  }
}

impl TryFrom<&str> for SourceType {
  type Error = rspack_error::Error;

  fn try_from(value: &str) -> Result<Self, Self::Error> {
    match value {
      "javascript" => Ok(Self::JavaScript),
      "css" => Ok(Self::Css),
      "wasm" => Ok(Self::Wasm),
      "asset" => Ok(Self::Asset),
      "expose" => Ok(Self::Expose),
      "remote" => Ok(Self::Remote),
      "share-init" => Ok(Self::ShareInit),
      "consume-shared" => Ok(Self::ConsumeShared),
      "unknown" => Ok(Self::Unknown),

      _ => {
        use rspack_error::error;
        Err(error!("invalid source type: {value}"))
      }
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ModuleType {
  Json,
  Css,
  CssModule,
  CssAuto,
  Js,
  JsDynamic,
  JsEsm,
  WasmSync,
  WasmAsync,
  AssetInline,
  AssetResource,
  AssetSource,
  Asset,
  Runtime,
  Remote,
  Fallback,
  ProvideShared,
  ConsumeShared,
  SelfReference,
  Custom(Ustr),
}

impl ModuleType {
  pub fn is_css_like(&self) -> bool {
    matches!(self, Self::Css | Self::CssModule | Self::CssAuto)
  }

  pub fn is_js_like(&self) -> bool {
    matches!(
      self,
      ModuleType::Js | ModuleType::JsEsm | ModuleType::JsDynamic
    )
  }

  pub fn is_asset_like(&self) -> bool {
    matches!(
      self,
      ModuleType::Asset | ModuleType::AssetInline | ModuleType::AssetResource
    )
  }

  pub fn is_wasm_like(&self) -> bool {
    matches!(self, ModuleType::WasmSync | ModuleType::WasmAsync)
  }

  pub fn is_js_auto(&self) -> bool {
    matches!(self, ModuleType::Js)
  }

  pub fn is_js_esm(&self) -> bool {
    matches!(self, ModuleType::JsEsm)
  }

  pub fn is_js_dynamic(&self) -> bool {
    matches!(self, ModuleType::JsDynamic)
  }

  /// Webpack arbitrary determines the binary type from [NormalModule.binary](https://github.com/webpack/webpack/blob/1f99ad6367f2b8a6ef17cce0e058f7a67fb7db18/lib/NormalModule.js#L302)
  pub fn is_binary(&self) -> bool {
    self.is_asset_like() || self.is_wasm_like()
  }

  pub fn as_str(&self) -> &str {
    match self {
      ModuleType::Js => "javascript/auto",
      ModuleType::JsEsm => "javascript/esm",
      ModuleType::JsDynamic => "javascript/dynamic",

      ModuleType::Css => "css",
      ModuleType::CssModule => "css/module",
      ModuleType::CssAuto => "css/auto",

      ModuleType::Json => "json",

      ModuleType::WasmSync => "webassembly/sync",
      ModuleType::WasmAsync => "webassembly/async",

      ModuleType::Asset => "asset",
      ModuleType::AssetSource => "asset/source",
      ModuleType::AssetResource => "asset/resource",
      ModuleType::AssetInline => "asset/inline",
      ModuleType::Runtime => "runtime",
      ModuleType::Remote => "remote-module",
      ModuleType::Fallback => "fallback-module",
      ModuleType::ProvideShared => "provide-module",
      ModuleType::ConsumeShared => "consume-shared-module",
      ModuleType::SelfReference => "self-reference-module",

      ModuleType::Custom(custom) => custom,
    }
  }
}

impl fmt::Display for ModuleType {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    write!(f, "{}", self.as_str(),)
  }
}

impl From<&str> for ModuleType {
  fn from(value: &str) -> Self {
    match value {
      "mjs" => Self::JsEsm,
      "cjs" => Self::JsDynamic,
      "js" | "javascript" | "js/auto" | "javascript/auto" => Self::Js,
      "js/dynamic" | "javascript/dynamic" => Self::JsDynamic,
      "js/esm" | "javascript/esm" => Self::JsEsm,

      "css" => Self::Css,
      "css/module" => Self::CssModule,
      "css/auto" => Self::CssAuto,

      "json" => Self::Json,

      "webassembly/sync" => Self::WasmSync,
      "webassembly/async" => Self::WasmAsync,

      "asset" => Self::Asset,
      "asset/resource" => Self::AssetResource,
      "asset/source" => Self::AssetSource,
      "asset/inline" => Self::AssetInline,

      custom => Self::Custom(custom.into()),
    }
  }
}

pub type ChunkByUkey = Database<Chunk>;
pub type ChunkGroupByUkey = Database<ChunkGroup>;
pub(crate) type SharedPluginDriver = Arc<PluginDriver>;

pub fn get_chunk_group_from_ukey<'a>(
  ukey: &ChunkGroupUkey,
  chunk_group_by_ukey: &'a ChunkGroupByUkey,
) -> Option<&'a ChunkGroup> {
  if chunk_group_by_ukey.contains(ukey) {
    Some(chunk_group_by_ukey.expect_get(ukey))
  } else {
    None
  }
}

pub fn get_chunk_from_ukey<'a>(
  ukey: &ChunkUkey,
  chunk_by_ukey: &'a ChunkByUkey,
) -> Option<&'a Chunk> {
  if chunk_by_ukey.contains(ukey) {
    Some(chunk_by_ukey.expect_get(ukey))
  } else {
    None
  }
}

pub fn get_mut_chunk_from_ukey<'a>(
  ukey: &ChunkUkey,
  chunk_by_ukey: &'a mut ChunkByUkey,
) -> Option<&'a mut Chunk> {
  if chunk_by_ukey.contains(ukey) {
    Some(chunk_by_ukey.expect_get_mut(ukey))
  } else {
    None
  }
}
