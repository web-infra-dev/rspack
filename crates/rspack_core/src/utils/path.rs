use std::path::{Component, Path};
use sugar_path::{self, PathSugar};
pub fn normalize_path(path: &str, root: &str) -> String {
  let res = Path::new(&root)
    .relative(Path::new(&path))
    .to_string_lossy()
    .to_string();

  res
}

pub fn uri_to_chunk_name(root: &str, uri: &str) -> String {
  println!("root {:?}", root);
  let path = Path::new(uri);
  let mut relatived = Path::new(root).relative(path);
  let ext = relatived
    .extension()
    .and_then(|ext| ext.to_str())
    .unwrap_or("")
    .to_string();
  relatived.set_extension("");
  let mut name = relatived
    .components()
    .filter(|com| matches!(com, Component::Normal(_)))
    .filter_map(|seg| seg.as_os_str().to_str())
    .intersperse("_")
    .fold(String::new(), |mut acc, seg| {
      acc.push_str(seg);
      acc
    });
  name.push('_');
  name.push_str(&ext);
  name
}
