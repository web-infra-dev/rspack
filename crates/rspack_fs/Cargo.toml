[package]
edition    = "2021"
license    = "MIT"
name       = "rspack_fs"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

[features]
async        = ["dep:futures", "dep:tokio"]
default      = ["native"]
native       = []
rspack-error = ["dep:rspack_error"]


[dependencies.rspack_error]
optional = true
path     = "../rspack_error"

[dependencies.futures]
optional  = true
workspace = true

[dependencies.tokio]
optional  = true
workspace = true
