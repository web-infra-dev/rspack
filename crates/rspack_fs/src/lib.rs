mod fs;
pub use fs::FileSystem;
mod read;
pub use read::ReadableFileSystem;

mod write;
pub use write::WritableFileSystem;
mod intermediate;
pub use intermediate::{
  IntermediateFileSystem, IntermediateFileSystemExtras, ReadStream, WriteStream,
};

mod file_metadata;
pub use file_metadata::FileMetadata;

mod macros;

mod native_fs;
pub use native_fs::{NativeFileSystem, NativeReadStream, NativeWriteStream};

mod memory_fs;
pub use memory_fs::{MemoryFileSystem, MemoryReadStream, MemoryWriteStream};

mod error;
pub use error::{Error, Result};
