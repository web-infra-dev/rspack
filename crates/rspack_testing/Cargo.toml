[package]
edition    = "2021"
license    = "MIT"
name       = "rspack_testing"
repository = "https://github.com/web-infra-dev/rspack"
version    = "0.1.0"

[[bin]]
bench = false
name  = "gen_test_config_schema"
path  = "./bin/gen_test_config_schema.rs"
test  = false

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rspack_binding_options                  = { path = "../rspack_binding_options" }
rspack_core                             = { path = "../rspack_core" }
rspack_error                            = { path = "../rspack_error" }
rspack_fs                               = { path = "../rspack_fs", features = ["async", "rspack-error"] }
rspack_ids                              = { path = "../rspack_ids" }
rspack_loader_runner                    = { path = "../rspack_loader_runner" }
rspack_loader_sass                      = { path = "../rspack_loader_sass" }
rspack_loader_swc                       = { path = "../rspack_loader_swc" }
rspack_plugin_asset                     = { path = "../rspack_plugin_asset" }
rspack_plugin_css                       = { path = "../rspack_plugin_css" }
rspack_plugin_dev_friendly_split_chunks = { path = "../rspack_plugin_dev_friendly_split_chunks" }
rspack_plugin_devtool                   = { path = "../rspack_plugin_devtool" }
rspack_plugin_entry                     = { path = "../rspack_plugin_entry" }
rspack_plugin_externals                 = { path = "../rspack_plugin_externals" }
rspack_plugin_hmr                       = { path = "../rspack_plugin_hmr" }
rspack_plugin_html                      = { path = "../rspack_plugin_html", features = ["testing"] }
rspack_plugin_javascript                = { path = "../rspack_plugin_javascript" }
rspack_plugin_json                      = { path = "../rspack_plugin_json" }
rspack_plugin_library                   = { path = "../rspack_plugin_library" }
rspack_plugin_remove_empty_chunks       = { path = "../rspack_plugin_remove_empty_chunks" }
rspack_plugin_runtime                   = { path = "../rspack_plugin_runtime" }
rspack_plugin_warn_sensitive_module     = { path = "../rspack_plugin_warn_sensitive_module" }
rspack_plugin_wasm                      = { path = "../rspack_plugin_wasm" }
rspack_regex                            = { path = "../rspack_regex" }
rspack_tracing                          = { path = "../rspack_tracing" }

async-trait    = { workspace = true }
cargo-rst      = { path = "../cargo-rst" }
schemars       = { workspace = true }
serde          = { workspace = true }
serde_json     = { workspace = true }
swc_core       = { workspace = true }
testing_macros = { workspace = true, features = [] }
tokio          = { workspace = true, features = ["rt", "rt-multi-thread", "macros", "test-util", "parking_lot"] }

insta     = { version = "1.31.0", features = ["json"] }
itertools = { workspace = true }
