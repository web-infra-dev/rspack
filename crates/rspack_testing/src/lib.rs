mod eval_raw;
mod loader;
mod run_fixture;
mod test_config;
pub use eval_raw::evaluate_to_json;
pub use run_fixture::{
  apply_from_fixture, test_fixture, test_fixture_css, test_fixture_css_modules, test_fixture_html,
  test_fixture_insta, test_fixture_js, test_rebuild_fixture,
};
pub use test_config::TestConfig;
pub use testing_macros::{self, fixture};
