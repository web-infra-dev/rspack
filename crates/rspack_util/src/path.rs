use std::path::{Component, Components, Path, PathBuf};

fn normalize(parts: Components, allow_above_root: bool) -> Vec<Component> {
  let mut res = vec![];
  for p in parts {
    match &p {
      Component::CurDir => (),
      Component::ParentDir => {
        if !matches!(
          res.last(),
          Some(Component::ParentDir) | Some(Component::RootDir)
        ) {
          res.pop();
        } else if allow_above_root {
          res.push(p);
        }
      }
      _ => res.push(p),
    }
  }
  res
}

pub fn relative(from: &Path, to: &Path) -> PathBuf {
  if from == to {
    return PathBuf::new();
  }

  let is_from_absolute = matches!(from.components().next(), Some(Component::RootDir));
  let is_to_absolute = matches!(to.components().next(), Some(Component::RootDir));

  // At this point the path should be resolved to a full absolute path, but
  // handle relative paths to be safe

  // Normalize the path
  let from = normalize(from.components(), !is_from_absolute);
  let to = normalize(to.components(), !is_to_absolute);

  let mut from_iter = from.iter();
  let mut to_iter = to.iter();
  let mut common_parts = 0;

  let from_remain = loop {
    match (from_iter.next(), to_iter.next()) {
      (Some(from_part), Some(to_part)) if from_part == to_part => {
        common_parts += 1;
      }
      (None, _) => break 0,
      _ => break from_iter.count() + 1,
    }
  };

  let mut result = PathBuf::new();
  for _ in 0..from_remain {
    result.push("..");
  }

  let to_iter = to.into_iter().skip(common_parts);
  for part in to_iter {
    result.push(part.as_os_str());
  }

  result
}

#[cfg(test)]
mod test {
  use std::path::Path;

  use super::*;

  #[cfg(not(target_os = "windows"))]
  #[test]
  fn test_posix() {
    let test_cases = vec![
      ("/var/lib", "/var", ".."),
      ("/var/lib", "/bin", "../../bin"),
      ("/var/lib", "/var/lib", ""),
      ("/var/lib", "/var/apache", "../apache"),
      ("/var/", "/var/lib", "lib"),
      ("/", "/var/lib", "var/lib"),
      (
        "/foo/test",
        "/foo/test/bar/package.json",
        "bar/package.json",
      ),
      ("/Users/a/web/b/test/mails", "/Users/a/web/b", "../.."),
      ("/foo/bar/baz-quux", "/foo/bar/baz", "../baz"),
      ("/foo/bar/baz", "/foo/bar/baz-quux", "../baz-quux"),
      ("/baz-quux", "/baz", "../baz"),
      ("/baz", "/baz-quux", "../baz-quux"),
      ("/page1/page2/foo", "/", "../../.."),
      // Fix https://github.com/web-infra-dev/rspack/issues/8219
      ("/", "/../maps/main.js.map", "maps/main.js.map"),
    ];

    for (from, to, expected) in test_cases {
      let actual = relative(Path::new(from), Path::new(to))
        .to_string_lossy()
        .to_string();
      assert_eq!(actual, expected.to_string());
    }
  }

  #[cfg(target_os = "windows")]
  #[test]
  fn test_win32() {
    let test_cases = vec![
      ("c:/blah\\blah", "d:/games", "d:\\games"),
      ("c:/aaaa/bbbb", "c:/aaaa", ".."),
      ("c:/aaaa/bbbb", "c:/cccc", "..\\..\\cccc"),
      ("c:/aaaa/bbbb", "c:/aaaa/bbbb", ""),
      ("c:/aaaa/bbbb", "c:/aaaa/cccc", "..\\cccc"),
      ("c:/aaaa/", "c:/aaaa/cccc", "cccc"),
      ("c:/", "c:\\aaaa\\bbbb", "aaaa\\bbbb"),
      ("c:/aaaa/bbbb", "d:\\", "d:\\"),
      ("c:/AaAa/bbbb", "c:/aaaa/bbbb", ""),
      ("c:/aaaaa/", "c:/aaaa/cccc", "..\\aaaa\\cccc"),
      ("C:\\foo\\bar\\baz\\quux", "C:\\", "..\\..\\..\\.."),
      (
        "C:\\foo\\test",
        "C:\\foo\\test\\bar\\package.json",
        "bar\\package.json",
      ),
      ("C:\\foo\\bar\\baz-quux", "C:\\foo\\bar\\baz", "..\\baz"),
      (
        "C:\\foo\\bar\\baz",
        "C:\\foo\\bar\\baz-quux",
        "..\\baz-quux",
      ),
      ("\\\\foo\\bar", "\\\\foo\\bar\\baz", "baz"),
      ("\\\\foo\\bar\\baz", "\\\\foo\\bar", ".."),
      ("\\\\foo\\bar\\baz-quux", "\\\\foo\\bar\\baz", "..\\baz"),
      (
        "\\\\foo\\bar\\baz",
        "\\\\foo\\bar\\baz-quux",
        "..\\baz-quux",
      ),
      ("C:\\baz-quux", "C:\\baz", "..\\baz"),
      ("C:\\baz", "C:\\baz-quux", "..\\baz-quux"),
      ("\\\\foo\\baz-quux", "\\\\foo\\baz", "..\\baz"),
      ("\\\\foo\\baz", "\\\\foo\\baz-quux", "..\\baz-quux"),
      ("C:\\baz", "\\\\foo\\bar\\baz", "\\\\foo\\bar\\baz"),
      ("\\\\foo\\bar\\baz", "C:\\baz", "C:\\baz"),
    ];

    for (from, to, expected) in test_cases {
      let actual = relative(Path::new(from), Path::new(to))
        .to_string_lossy()
        .to_string();
      assert_eq!(actual, expected.to_string());
    }
  }
}
