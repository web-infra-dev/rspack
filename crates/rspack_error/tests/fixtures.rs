use std::path::PathBuf;

use insta::Settings;
use rspack_core::{Compiler, Stats};
use rspack_fs::AsyncNativeFileSystem;
use rspack_testing::{apply_from_fixture, fixture};
use rspack_tracing::enable_tracing_by_env;

#[tokio::main]
pub async fn test_fixture<F: FnOnce(&Stats, Settings) -> rspack_error::Result<()>>(
  fixture_path: &PathBuf,
  f: F,
) -> rspack_error::Result<()> {
  let (options, plugins) = apply_from_fixture(fixture_path);
  let mut compiler = Compiler::new(options, plugins, AsyncNativeFileSystem);

  compiler
    .build()
    .await
    .unwrap_or_else(|_| panic!("failed to compile in fixtrue {fixture_path:?}"));
  let stats = compiler.compilation.get_stats();
  let mut settings = Settings::clone_current();
  settings.remove_snapshot_suffix();
  settings.set_snapshot_path(fixture_path);
  f(&stats, settings)
}

#[fixture("tests/fixtures/*", exclude("export_star_error"))]
fn custom(fixture_path: PathBuf) {
  enable_tracing_by_env();
  test_fixture(&fixture_path, |stats, settings| {
    let dirname = fixture_path
      .components()
      .last()
      .expect("TODO:")
      .as_os_str()
      .to_str()
      .expect("TODO:")
      .to_owned();
    settings.bind(|| {
      insta::assert_snapshot!(
        dirname.as_str(),
        stats.emit_diagnostics_string(false).expect("TODO:"),
        dirname.as_str()
      );
    });
    Ok(())
  })
  .expect("TODO:");
}

/// In concurrent scenario the file resolve order can't be guaranteed.
/// So we need to sort the diagnostic before writing into snapshot
#[fixture("tests/out_of_order/*")]
fn out_of_order(fixture_path: PathBuf) {
  test_fixture(&fixture_path, |stats, settings| {
    let dirname = fixture_path
      .components()
      .last()
      .expect("TODO:")
      .as_os_str()
      .to_str()
      .expect("TODO:")
      .to_owned();
    settings.bind(|| {
      insta::assert_snapshot!(
        dirname.as_str(),
        stats.emit_diagnostics_string(true).expect("TODO:"),
        dirname.as_str()
      );
    });
    Ok(())
  })
  .expect("TODO:");
}
