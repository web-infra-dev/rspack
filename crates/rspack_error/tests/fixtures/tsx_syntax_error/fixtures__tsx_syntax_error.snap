---
source: crates/rspack_error/tests/fixtures.rs
expression: tsx_syntax_error
---
error[tsx]: TSX parsing error
  ┌─ tests/fixtures/tsx_syntax_error/index.tsx:1:10
  │
1 │ let c = <test>/test>;
  │          ^^^^ Unexpected token `test`. Expected jsx identifier
2 │ 


